### PR TITLE
Fix dependently-typed functions with type reference union member types

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/43012

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java [hunk 1]
```diff
@@ -986,6 +986,7 @@ private BType getMatchingTypeForInferrableType(BType originalType, BType expType
 
         for (BType inferableType : inferableTypes) {
             for (BType expectedType : expectedTypes) {
+                expectedType = Types.getImpliedType(expectedType);
                 if (Types.getImpliedType(inferableType).tag == Types.getImpliedType(expectedType).tag) {
```
Comment (LakshanWeerasinghe) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java:990]:
```suggestion
                if (Types.getImpliedType(inferableType).tag == expectedType.tag) {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/javainterop/dependently_typed_functions_test_negative.bal [hunk 2]
```diff
@@ -370,3 +370,15 @@ function testDependentlyTypedFunctionWithInferredArgForParamOfTypeReferenceTypeN
...
+function getArrayInferred(typedesc<anydata> td = <>) returns td[] = @java:Method {
+    'class: "org.ballerinalang.nativeimpl.jvm.tests.VariableReturnType",
+    name: "getArray",
+    paramTypes: ["io.ballerina.runtime.api.values.BTypedesc"]
+} external;
+
+type AnyArray any[];
+
+function testDependentlyTypedFunctionWithTypeReferenceTypeNegative() returns error? {
+    AnyArray _ = check getArrayInferred();
```
Comment (SasinduDilshara) [inline tests/jballerina-unit-test/src/test/resources/test-src/javainterop/dependently_typed_functions_test_negative.bal:383]: Shall we add a test with union types as the expected type for confirm the behaviour.

ex:- `type C A|B`

and C as the expected type

Comment (MaryamZi) [reply tests/jballerina-unit-test/src/test/resources/test-src/javainterop/dependently_typed_functions_test_negative.bal:383]: This issue arises when a member of the union is a type reference type, that is covered with the usage of check when the expected type is a type reference type - effectively `IntArray|error`.

**Review by SasinduDilshara — APPROVED** [review]

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java [hunk 3]
```diff
@@ -986,7 +986,8 @@ private BType getMatchingTypeForInferrableType(BType originalType, BType expType
 
         for (BType inferableType : inferableTypes) {
             for (BType expectedType : expectedTypes) {
-                if (Types.getImpliedType(inferableType).tag == Types.getImpliedType(expectedType).tag) {
+                expectedType = Types.getImpliedType(expectedType);
+                if (Types.getImpliedType(inferableType).tag == expectedType.tag) {
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java:990]: Is this correct? When creating the `inferableTypes` and `expectedTypes` lists, don't we need to add the original types(by doing the `getImpliedType`  when creating the lists)?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/Unifier.java:990]: I was considering the same. But we seem to have worked with implied types even in the non-union case, so decided to do the same in the union case also.

**Pull request merged**